src/components/heartrate/Ppg.cpp

```diff
@@ -2,98 +2,242 @@
+  }
+  dataIndex = dataLength - overlapWindow;
+  return hr;
+}
+// Pass init == true to reset spectral averaging.
+// Returns -1 (Reset Acquisition), 0 (Unable to obtain HR) or HR (BPM).
+int Ppg::ProcessHeartRate(bool init) {
+  std::copy(dataHRS.begin(), dataHRS.end(), vReal.begin());
+  Detrend(vReal.data(), dataLength);
+  Filter30to240(vReal.data(), dataLength);
+  // Compute in place power spectrum
+  vImag.fill(0.0f);
+  FFT.Windowing(vReal.data(), static_cast<uint16_t>(dataLength), FFT_WIN_TYP_HANN, FFT_FORWARD);
+  FFT.Compute(vReal.data(), vImag.data(), static_cast<uint16_t>(dataLength), FFT_FORWARD);
+  FFT.ComplexToMagnitude(vReal.data(), vImag.data(), static_cast<uint16_t>(dataLength));
+  SpectrumAverage(vReal.data(), spectrum.data(), spectrum.size(), init);
+  peakLocation = 0.0f;
+  float threshold = peakDetectionThreshold;
+  float peakWidth = 0.0f;
+  int specLen = spectrum.size();
+  float max = Max(spectrum.data(), hrROIbegin, hrROIend);
+  float signalToNoiseRatio = SignalToNoise(spectrum.data(), max, hrROIbegin, hrROIend);
+  if (signalToNoiseRatio > signalToNoiseThreshold && spectrum.at(0) < dcThreshold) {
+    threshold *= max;
+    // Reuse VImag for interpolation x values passed to PeakSearch
+    for (int i = 0; i < dataLength; i++) {
+      vImag[i] = i;
     }
+    peakLocation = PeakSearch(vImag.data(),
+                              spectrum.data(),
+                              specLen,
+                              static_cast<float>(hrROIbegin),
+                              static_cast<float>(hrROIend),
+                              threshold,
+                              peakWidth);
+    peakLocation *= freqResolution;
+  }
+  // Peak too wide? (broad spectrum noise or large, rapid HR change)
+  if (peakWidth > maxPeakWidth)
+    peakLocation = 0.0f;
+  // Check HR limits
+  if (peakLocation < minHR || peakLocation > maxHR)
+    peakLocation = 0.0f;
+  // Reset spectral averaging if bad reading
+  if (peakLocation == 0.0f)
+    resetSpectralAvg = true;
+  // Set the ambient light threshold and return HR in BPM
+  alsThreshold = static_cast<uint32_t>(alsValue * alsFactor);
+  peakLocation = HeartRateAverage(peakLocation);
+  lastPeakLocation = peakLocation;
+  // Restart acq if averaging buffer zeroed out
+  if (peakLocation == 0.0f) {
     return -1;
+  } else {
+    return static_cast<int>(roundf(peakLocation * 60.0f));
   }
 }
 
-Ppg::Ppg()
-  : hpf {0.87033078, -1.74066156, 0.87033078, -1.72377617, 0.75754694},
-    agc {20, 0.971, 2},
-    lpf {0.11595249, 0.23190498, 0.11595249, -0.72168143, 0.18549138} {
+float Ppg::PeakSearch(const float* xVals, const float* yVals, int length, float start, float end, float th, float& width) {
+  int peaks = 0;
+  bool enabled = false;
+  float v0, v1, v2, minBin = 0.0f, maxBin = 0.0f, peakCenter = 0.0f;
+  v0 = Interpolation::CatmullSpline(xVals, yVals, length, start - 0.01f);
+  v1 = Interpolation::CatmullSpline(xVals, yVals, length, start);
```

Comment:
Prefer initializing while declaring.
```suggestion
  float minBin = 0.0f, maxBin = 0.0f, peakCenter = 0.0f;
  float v0 = Interpolation::CatmullSpline(xVals, yVals, length, start - 0.01f);
  float v1 = Interpolation::CatmullSpline(xVals, yVals, length, start);
```

Reply:
Fixed in recent commits. Clang-tidy comes back clean except for some complaints about short variable names.